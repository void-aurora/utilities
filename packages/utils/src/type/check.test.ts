import { describe, test, expect } from 'vitest';
import {
  hasOwn,
  isArray,
  isAssigned,
  isBigInt,
  isBoolean,
  isDate,
  isDefined,
  isFunction,
  isIterable,
  isMap,
  isNull,
  isNullish,
  isNullOrUndefined,
  isNumber,
  isObject,
  isLiteralObject,
  isPrimitives,
  isPromise,
  isSet,
  isString,
  isSymbol,
  isUndefined,
  toRawType,
  toTypeString,
  isPlainObject,
} from './check';

class Foobar {
  foo: string;
  bar: string;
  constructor(foo: string, bar: string) {
    this.foo = foo;
    this.bar = bar;
  }
  getFoo() {
    return this.foo;
  }
  getBar() {
    return this.bar;
  }
  get [Symbol.toStringTag]() {
    return 'Foobar';
  }
  *[Symbol.iterator](): Iterator<string> {
    yield 'foo';
    yield 'bar';
  }
  toString() {
    return this.foo + this.bar;
  }
  hasOwnProperty(key: string) {
    return true;
  }
}

describe('type/check', () => {
  test(`basic`, () => {
    expect(isNull(null)).toBe(true);
    expect(isUndefined(undefined)).toBe(true);
    expect(isNullOrUndefined(void 0)).toBe(true);
    expect(isAssigned({})).toBe(true);
    expect(isString('foobar')).toBe(true);
    expect(isNumber(123)).toBe(true);
    expect(isBoolean(false)).toBe(true);
    expect(isSymbol(Symbol('foobar'))).toBe(true);
    expect(isBigInt(BigInt('321'))).toBe(true);
    expect(isPrimitives(456)).toBe(true);
    expect(isFunction(() => false)).toBe(true);
    expect(isObject({})).toBe(true);
    expect(isLiteralObject({})).toBe(true);
    expect(isIterable([])).toBe(true);
    expect(isArray([])).toBe(true);
    expect(isMap(new Map())).toBe(true);
    expect(isSet(new Set())).toBe(true);
    expect(
      isPromise(
        new Promise((resolve, reject) => setTimeout(() => resolve(1), 100)),
      ),
    ).toBe(true);
    expect(isDate(new Date())).toBe(true);
    expect(toTypeString(123)).toBe('[object Number]');
    expect(toRawType(123)).toBe('Number');
  });

  test(`snapshot`, () => {
    const checkers = Object.entries({
      isNull,
      isUndefined,
      isNullOrUndefined,
      isAssigned,
      isString,
      isNumber,
      isBoolean,
      isSymbol,
      isBigInt,
      isPrimitives,
      isFunction,
      isObject,
      isLiteralObject,
      isPlainObject,
      isIterable,
      isArray,
      isMap,
      isSet,
      isPromise,
      isDate,
      toTypeString,
      toRawType,
    } as Record<string, (v: unknown) => any>);

    function functionNormal(a: string, b: string): string {
      return a + b;
    }
    const functionArrow = (a: number, b: number) => a + b;

    const promiseFunction = () =>
      new Promise((resolve, reject) => {
        setTimeout(() => resolve(1), 100);
      });
    const asyncFunction = async () => {
      return 1;
    };

    const values = Object.entries({
      'primitives.undefined': undefined,
      'primitives.string': '11',
      'primitives.string.empty': '',
      'primitives.number': 11,
      'primitives.number.NaN': NaN,
      'primitives.number.NEGATIVE_INFINITY': Number.NEGATIVE_INFINITY,
      'primitives.number.POSITIVE_INFINITY': Number.POSITIVE_INFINITY,
      'primitives.boolean.true': true,
      'primitives.boolean.false': false,
      'primitives.symbol': Symbol('foobar'),
      'primitives.bigint': 11n,
      functionNormal,
      functionArrow,
      'object.literal': { foobar: 'foobar' },
      'object.plain': Object.create(null),
      'object.null': null,
      classFoobar: Foobar,
      instanceFoobar: new Foobar('a', 'b'),
      instanceArray: [0, 1, 2, 3, 4],
      instanceMap: new Map(),
      instanceSet: new Set(),
      instancePromise: promiseFunction(),
      instanceAsync: asyncFunction(),
      instanceDate: new Date(),
      apiString: String,
      apiNumber: Number,
      apiObject: Object,
      apiArray: Array,
      apiMath: Math,
      apiDate: Date,
    } as Record<string, any>);

    const result = Object.fromEntries(
      checkers.map(([checkerName, checker]) => [
        checkerName,
        Object.fromEntries(
          values.map(([valueName, value]) => [valueName, checker(value)]),
        ),
      ]),
    );

    expect(result).toMatchFileSnapshot(`${__filename}.snap`);
  });

  test('alias', () => {
    expect(isDefined).toBe(isAssigned);
    expect(isNullish).toBe(isNullOrUndefined);
  });

  test('hasOwn()', () => {
    const foobar = new Foobar('apple', 'banana');
    expect(hasOwn(foobar, 'foo')).toBe(true);
    expect(hasOwn(foobar, 'bar')).toBe(true);
    expect(hasOwn(foobar, 'toString')).toBe(false);
    expect(hasOwn(foobar, 'hasOwnProperty')).toBe(false);
    expect(hasOwn(foobar, 'valueOf')).toBe(false);
  });

  test('test type define isObject()', () => {
    const obj: unknown = {};
    const s = Symbol('foobar');
    if (isObject(obj)) {
      obj[1] = 'a';
      obj['a'] = 1;
      obj[s] = { foo: 'bar' };
    }
    expect(obj).toEqual({ [1]: 'a', a: 1, [s]: { foo: 'bar' } });
  });
});
