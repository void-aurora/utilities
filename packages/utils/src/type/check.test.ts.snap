{
  "isArray": {
    "apiArray": false,
    "apiDate": false,
    "apiMath": false,
    "apiNumber": false,
    "apiObject": false,
    "apiString": false,
    "classFoobar": false,
    "functionArrow": false,
    "functionNormal": false,
    "instanceArray": true,
    "instanceAsync": false,
    "instanceDate": false,
    "instanceFoobar": false,
    "instanceMap": false,
    "instancePromise": false,
    "instanceSet": false,
    "object": false,
    "object.null": false,
    "primitives.bigint": false,
    "primitives.boolean.false": false,
    "primitives.boolean.true": false,
    "primitives.number": false,
    "primitives.number.NEGATIVE_INFINITY": false,
    "primitives.number.NaN": false,
    "primitives.number.POSITIVE_INFINITY": false,
    "primitives.string": false,
    "primitives.string.empty": false,
    "primitives.symbol": false,
    "primitives.undefined": false,
  },
  "isAssigned": {
    "apiArray": true,
    "apiDate": true,
    "apiMath": true,
    "apiNumber": true,
    "apiObject": true,
    "apiString": true,
    "classFoobar": true,
    "functionArrow": true,
    "functionNormal": true,
    "instanceArray": true,
    "instanceAsync": true,
    "instanceDate": true,
    "instanceFoobar": true,
    "instanceMap": true,
    "instancePromise": true,
    "instanceSet": true,
    "object": true,
    "object.null": false,
    "primitives.bigint": true,
    "primitives.boolean.false": true,
    "primitives.boolean.true": true,
    "primitives.number": true,
    "primitives.number.NEGATIVE_INFINITY": true,
    "primitives.number.NaN": true,
    "primitives.number.POSITIVE_INFINITY": true,
    "primitives.string": true,
    "primitives.string.empty": true,
    "primitives.symbol": true,
    "primitives.undefined": false,
  },
  "isBigInt": {
    "apiArray": false,
    "apiDate": false,
    "apiMath": false,
    "apiNumber": false,
    "apiObject": false,
    "apiString": false,
    "classFoobar": false,
    "functionArrow": false,
    "functionNormal": false,
    "instanceArray": false,
    "instanceAsync": false,
    "instanceDate": false,
    "instanceFoobar": false,
    "instanceMap": false,
    "instancePromise": false,
    "instanceSet": false,
    "object": false,
    "object.null": false,
    "primitives.bigint": true,
    "primitives.boolean.false": false,
    "primitives.boolean.true": false,
    "primitives.number": false,
    "primitives.number.NEGATIVE_INFINITY": false,
    "primitives.number.NaN": false,
    "primitives.number.POSITIVE_INFINITY": false,
    "primitives.string": false,
    "primitives.string.empty": false,
    "primitives.symbol": false,
    "primitives.undefined": false,
  },
  "isBoolean": {
    "apiArray": false,
    "apiDate": false,
    "apiMath": false,
    "apiNumber": false,
    "apiObject": false,
    "apiString": false,
    "classFoobar": false,
    "functionArrow": false,
    "functionNormal": false,
    "instanceArray": false,
    "instanceAsync": false,
    "instanceDate": false,
    "instanceFoobar": false,
    "instanceMap": false,
    "instancePromise": false,
    "instanceSet": false,
    "object": false,
    "object.null": false,
    "primitives.bigint": false,
    "primitives.boolean.false": true,
    "primitives.boolean.true": true,
    "primitives.number": false,
    "primitives.number.NEGATIVE_INFINITY": false,
    "primitives.number.NaN": false,
    "primitives.number.POSITIVE_INFINITY": false,
    "primitives.string": false,
    "primitives.string.empty": false,
    "primitives.symbol": false,
    "primitives.undefined": false,
  },
  "isDate": {
    "apiArray": false,
    "apiDate": false,
    "apiMath": false,
    "apiNumber": false,
    "apiObject": false,
    "apiString": false,
    "classFoobar": false,
    "functionArrow": false,
    "functionNormal": false,
    "instanceArray": false,
    "instanceAsync": false,
    "instanceDate": true,
    "instanceFoobar": false,
    "instanceMap": false,
    "instancePromise": false,
    "instanceSet": false,
    "object": false,
    "object.null": false,
    "primitives.bigint": false,
    "primitives.boolean.false": false,
    "primitives.boolean.true": false,
    "primitives.number": false,
    "primitives.number.NEGATIVE_INFINITY": false,
    "primitives.number.NaN": false,
    "primitives.number.POSITIVE_INFINITY": false,
    "primitives.string": false,
    "primitives.string.empty": false,
    "primitives.symbol": false,
    "primitives.undefined": false,
  },
  "isFunction": {
    "apiArray": true,
    "apiDate": true,
    "apiMath": false,
    "apiNumber": true,
    "apiObject": true,
    "apiString": true,
    "classFoobar": true,
    "functionArrow": true,
    "functionNormal": true,
    "instanceArray": false,
    "instanceAsync": false,
    "instanceDate": false,
    "instanceFoobar": false,
    "instanceMap": false,
    "instancePromise": false,
    "instanceSet": false,
    "object": false,
    "object.null": false,
    "primitives.bigint": false,
    "primitives.boolean.false": false,
    "primitives.boolean.true": false,
    "primitives.number": false,
    "primitives.number.NEGATIVE_INFINITY": false,
    "primitives.number.NaN": false,
    "primitives.number.POSITIVE_INFINITY": false,
    "primitives.string": false,
    "primitives.string.empty": false,
    "primitives.symbol": false,
    "primitives.undefined": false,
  },
  "isIterable": {
    "apiArray": false,
    "apiDate": false,
    "apiMath": false,
    "apiNumber": false,
    "apiObject": false,
    "apiString": false,
    "classFoobar": false,
    "functionArrow": false,
    "functionNormal": false,
    "instanceArray": true,
    "instanceAsync": false,
    "instanceDate": false,
    "instanceFoobar": true,
    "instanceMap": true,
    "instancePromise": false,
    "instanceSet": true,
    "object": false,
    "object.null": false,
    "primitives.bigint": false,
    "primitives.boolean.false": false,
    "primitives.boolean.true": false,
    "primitives.number": false,
    "primitives.number.NEGATIVE_INFINITY": false,
    "primitives.number.NaN": false,
    "primitives.number.POSITIVE_INFINITY": false,
    "primitives.string": false,
    "primitives.string.empty": false,
    "primitives.symbol": false,
    "primitives.undefined": false,
  },
  "isMap": {
    "apiArray": false,
    "apiDate": false,
    "apiMath": false,
    "apiNumber": false,
    "apiObject": false,
    "apiString": false,
    "classFoobar": false,
    "functionArrow": false,
    "functionNormal": false,
    "instanceArray": false,
    "instanceAsync": false,
    "instanceDate": false,
    "instanceFoobar": false,
    "instanceMap": true,
    "instancePromise": false,
    "instanceSet": false,
    "object": false,
    "object.null": false,
    "primitives.bigint": false,
    "primitives.boolean.false": false,
    "primitives.boolean.true": false,
    "primitives.number": false,
    "primitives.number.NEGATIVE_INFINITY": false,
    "primitives.number.NaN": false,
    "primitives.number.POSITIVE_INFINITY": false,
    "primitives.string": false,
    "primitives.string.empty": false,
    "primitives.symbol": false,
    "primitives.undefined": false,
  },
  "isNull": {
    "apiArray": false,
    "apiDate": false,
    "apiMath": false,
    "apiNumber": false,
    "apiObject": false,
    "apiString": false,
    "classFoobar": false,
    "functionArrow": false,
    "functionNormal": false,
    "instanceArray": false,
    "instanceAsync": false,
    "instanceDate": false,
    "instanceFoobar": false,
    "instanceMap": false,
    "instancePromise": false,
    "instanceSet": false,
    "object": false,
    "object.null": true,
    "primitives.bigint": false,
    "primitives.boolean.false": false,
    "primitives.boolean.true": false,
    "primitives.number": false,
    "primitives.number.NEGATIVE_INFINITY": false,
    "primitives.number.NaN": false,
    "primitives.number.POSITIVE_INFINITY": false,
    "primitives.string": false,
    "primitives.string.empty": false,
    "primitives.symbol": false,
    "primitives.undefined": false,
  },
  "isNullOrUndefined": {
    "apiArray": false,
    "apiDate": false,
    "apiMath": false,
    "apiNumber": false,
    "apiObject": false,
    "apiString": false,
    "classFoobar": false,
    "functionArrow": false,
    "functionNormal": false,
    "instanceArray": false,
    "instanceAsync": false,
    "instanceDate": false,
    "instanceFoobar": false,
    "instanceMap": false,
    "instancePromise": false,
    "instanceSet": false,
    "object": false,
    "object.null": true,
    "primitives.bigint": false,
    "primitives.boolean.false": false,
    "primitives.boolean.true": false,
    "primitives.number": false,
    "primitives.number.NEGATIVE_INFINITY": false,
    "primitives.number.NaN": false,
    "primitives.number.POSITIVE_INFINITY": false,
    "primitives.string": false,
    "primitives.string.empty": false,
    "primitives.symbol": false,
    "primitives.undefined": true,
  },
  "isNumber": {
    "apiArray": false,
    "apiDate": false,
    "apiMath": false,
    "apiNumber": false,
    "apiObject": false,
    "apiString": false,
    "classFoobar": false,
    "functionArrow": false,
    "functionNormal": false,
    "instanceArray": false,
    "instanceAsync": false,
    "instanceDate": false,
    "instanceFoobar": false,
    "instanceMap": false,
    "instancePromise": false,
    "instanceSet": false,
    "object": false,
    "object.null": false,
    "primitives.bigint": false,
    "primitives.boolean.false": false,
    "primitives.boolean.true": false,
    "primitives.number": true,
    "primitives.number.NEGATIVE_INFINITY": false,
    "primitives.number.NaN": false,
    "primitives.number.POSITIVE_INFINITY": false,
    "primitives.string": false,
    "primitives.string.empty": false,
    "primitives.symbol": false,
    "primitives.undefined": false,
  },
  "isObject": {
    "apiArray": false,
    "apiDate": false,
    "apiMath": true,
    "apiNumber": false,
    "apiObject": false,
    "apiString": false,
    "classFoobar": false,
    "functionArrow": false,
    "functionNormal": false,
    "instanceArray": true,
    "instanceAsync": true,
    "instanceDate": true,
    "instanceFoobar": true,
    "instanceMap": true,
    "instancePromise": true,
    "instanceSet": true,
    "object": true,
    "object.null": false,
    "primitives.bigint": false,
    "primitives.boolean.false": false,
    "primitives.boolean.true": false,
    "primitives.number": false,
    "primitives.number.NEGATIVE_INFINITY": false,
    "primitives.number.NaN": false,
    "primitives.number.POSITIVE_INFINITY": false,
    "primitives.string": false,
    "primitives.string.empty": false,
    "primitives.symbol": false,
    "primitives.undefined": false,
  },
  "isPlainObject": {
    "apiArray": false,
    "apiDate": false,
    "apiMath": false,
    "apiNumber": false,
    "apiObject": false,
    "apiString": false,
    "classFoobar": false,
    "functionArrow": false,
    "functionNormal": false,
    "instanceArray": false,
    "instanceAsync": false,
    "instanceDate": false,
    "instanceFoobar": false,
    "instanceMap": false,
    "instancePromise": false,
    "instanceSet": false,
    "object": true,
    "object.null": false,
    "primitives.bigint": false,
    "primitives.boolean.false": false,
    "primitives.boolean.true": false,
    "primitives.number": false,
    "primitives.number.NEGATIVE_INFINITY": false,
    "primitives.number.NaN": false,
    "primitives.number.POSITIVE_INFINITY": false,
    "primitives.string": false,
    "primitives.string.empty": false,
    "primitives.symbol": false,
    "primitives.undefined": false,
  },
  "isPrimitives": {
    "apiArray": false,
    "apiDate": false,
    "apiMath": false,
    "apiNumber": false,
    "apiObject": false,
    "apiString": false,
    "classFoobar": false,
    "functionArrow": false,
    "functionNormal": false,
    "instanceArray": false,
    "instanceAsync": false,
    "instanceDate": false,
    "instanceFoobar": false,
    "instanceMap": false,
    "instancePromise": false,
    "instanceSet": false,
    "object": false,
    "object.null": true,
    "primitives.bigint": true,
    "primitives.boolean.false": true,
    "primitives.boolean.true": true,
    "primitives.number": true,
    "primitives.number.NEGATIVE_INFINITY": false,
    "primitives.number.NaN": false,
    "primitives.number.POSITIVE_INFINITY": false,
    "primitives.string": true,
    "primitives.string.empty": true,
    "primitives.symbol": true,
    "primitives.undefined": true,
  },
  "isPromise": {
    "apiArray": false,
    "apiDate": false,
    "apiMath": false,
    "apiNumber": false,
    "apiObject": false,
    "apiString": false,
    "classFoobar": false,
    "functionArrow": false,
    "functionNormal": false,
    "instanceArray": false,
    "instanceAsync": true,
    "instanceDate": false,
    "instanceFoobar": false,
    "instanceMap": false,
    "instancePromise": true,
    "instanceSet": false,
    "object": false,
    "object.null": false,
    "primitives.bigint": false,
    "primitives.boolean.false": false,
    "primitives.boolean.true": false,
    "primitives.number": false,
    "primitives.number.NEGATIVE_INFINITY": false,
    "primitives.number.NaN": false,
    "primitives.number.POSITIVE_INFINITY": false,
    "primitives.string": false,
    "primitives.string.empty": false,
    "primitives.symbol": false,
    "primitives.undefined": false,
  },
  "isSet": {
    "apiArray": false,
    "apiDate": false,
    "apiMath": false,
    "apiNumber": false,
    "apiObject": false,
    "apiString": false,
    "classFoobar": false,
    "functionArrow": false,
    "functionNormal": false,
    "instanceArray": false,
    "instanceAsync": false,
    "instanceDate": false,
    "instanceFoobar": false,
    "instanceMap": false,
    "instancePromise": false,
    "instanceSet": true,
    "object": false,
    "object.null": false,
    "primitives.bigint": false,
    "primitives.boolean.false": false,
    "primitives.boolean.true": false,
    "primitives.number": false,
    "primitives.number.NEGATIVE_INFINITY": false,
    "primitives.number.NaN": false,
    "primitives.number.POSITIVE_INFINITY": false,
    "primitives.string": false,
    "primitives.string.empty": false,
    "primitives.symbol": false,
    "primitives.undefined": false,
  },
  "isString": {
    "apiArray": false,
    "apiDate": false,
    "apiMath": false,
    "apiNumber": false,
    "apiObject": false,
    "apiString": false,
    "classFoobar": false,
    "functionArrow": false,
    "functionNormal": false,
    "instanceArray": false,
    "instanceAsync": false,
    "instanceDate": false,
    "instanceFoobar": false,
    "instanceMap": false,
    "instancePromise": false,
    "instanceSet": false,
    "object": false,
    "object.null": false,
    "primitives.bigint": false,
    "primitives.boolean.false": false,
    "primitives.boolean.true": false,
    "primitives.number": false,
    "primitives.number.NEGATIVE_INFINITY": false,
    "primitives.number.NaN": false,
    "primitives.number.POSITIVE_INFINITY": false,
    "primitives.string": true,
    "primitives.string.empty": true,
    "primitives.symbol": false,
    "primitives.undefined": false,
  },
  "isSymbol": {
    "apiArray": false,
    "apiDate": false,
    "apiMath": false,
    "apiNumber": false,
    "apiObject": false,
    "apiString": false,
    "classFoobar": false,
    "functionArrow": false,
    "functionNormal": false,
    "instanceArray": false,
    "instanceAsync": false,
    "instanceDate": false,
    "instanceFoobar": false,
    "instanceMap": false,
    "instancePromise": false,
    "instanceSet": false,
    "object": false,
    "object.null": false,
    "primitives.bigint": false,
    "primitives.boolean.false": false,
    "primitives.boolean.true": false,
    "primitives.number": false,
    "primitives.number.NEGATIVE_INFINITY": false,
    "primitives.number.NaN": false,
    "primitives.number.POSITIVE_INFINITY": false,
    "primitives.string": false,
    "primitives.string.empty": false,
    "primitives.symbol": true,
    "primitives.undefined": false,
  },
  "isUndefined": {
    "apiArray": false,
    "apiDate": false,
    "apiMath": false,
    "apiNumber": false,
    "apiObject": false,
    "apiString": false,
    "classFoobar": false,
    "functionArrow": false,
    "functionNormal": false,
    "instanceArray": false,
    "instanceAsync": false,
    "instanceDate": false,
    "instanceFoobar": false,
    "instanceMap": false,
    "instancePromise": false,
    "instanceSet": false,
    "object": false,
    "object.null": false,
    "primitives.bigint": false,
    "primitives.boolean.false": false,
    "primitives.boolean.true": false,
    "primitives.number": false,
    "primitives.number.NEGATIVE_INFINITY": false,
    "primitives.number.NaN": false,
    "primitives.number.POSITIVE_INFINITY": false,
    "primitives.string": false,
    "primitives.string.empty": false,
    "primitives.symbol": false,
    "primitives.undefined": true,
  },
  "toRawType": {
    "apiArray": "Function",
    "apiDate": "Function",
    "apiMath": "Math",
    "apiNumber": "Function",
    "apiObject": "Function",
    "apiString": "Function",
    "classFoobar": "Function",
    "functionArrow": "Function",
    "functionNormal": "Function",
    "instanceArray": "Array",
    "instanceAsync": "Promise",
    "instanceDate": "Date",
    "instanceFoobar": "Foobar",
    "instanceMap": "Map",
    "instancePromise": "Promise",
    "instanceSet": "Set",
    "object": "Object",
    "object.null": "Null",
    "primitives.bigint": "BigInt",
    "primitives.boolean.false": "Boolean",
    "primitives.boolean.true": "Boolean",
    "primitives.number": "Number",
    "primitives.number.NEGATIVE_INFINITY": "Number",
    "primitives.number.NaN": "Number",
    "primitives.number.POSITIVE_INFINITY": "Number",
    "primitives.string": "String",
    "primitives.string.empty": "String",
    "primitives.symbol": "Symbol",
    "primitives.undefined": "Undefined",
  },
  "toTypeString": {
    "apiArray": "[object Function]",
    "apiDate": "[object Function]",
    "apiMath": "[object Math]",
    "apiNumber": "[object Function]",
    "apiObject": "[object Function]",
    "apiString": "[object Function]",
    "classFoobar": "[object Function]",
    "functionArrow": "[object Function]",
    "functionNormal": "[object Function]",
    "instanceArray": "[object Array]",
    "instanceAsync": "[object Promise]",
    "instanceDate": "[object Date]",
    "instanceFoobar": "[object Foobar]",
    "instanceMap": "[object Map]",
    "instancePromise": "[object Promise]",
    "instanceSet": "[object Set]",
    "object": "[object Object]",
    "object.null": "[object Null]",
    "primitives.bigint": "[object BigInt]",
    "primitives.boolean.false": "[object Boolean]",
    "primitives.boolean.true": "[object Boolean]",
    "primitives.number": "[object Number]",
    "primitives.number.NEGATIVE_INFINITY": "[object Number]",
    "primitives.number.NaN": "[object Number]",
    "primitives.number.POSITIVE_INFINITY": "[object Number]",
    "primitives.string": "[object String]",
    "primitives.string.empty": "[object String]",
    "primitives.symbol": "[object Symbol]",
    "primitives.undefined": "[object Undefined]",
  },
}